A6-client-retries.md

```diff
@@ -429,9 +432,12 @@ The retry policy is transmitted to the client through the service config mechani
         "maxAttempts": number,
 
         // The original RPC will be sent immediately, but the maxAttempts-1 
-        // subsequent hedged RPCs will be sent at intervals of every hedgingDelayMs. 
+        // subsequent hedged RPCs will be sent at intervals of every hedgingDelay. 
         // Set this to 0 to immediately send all maxAttempts RPCs.
-        "hedgingDelayMs": number,
+        // hedgingDelay takes its form from:
+        // https://developers.google.com/protocol-buffers/docs/proto3#json
```

Comment:
https://developers.google.com/protocol-buffers/docs/proto3#json doesn't adequately describe the "Duration" format.  It just gives 2 trivial examples.

For example: is "s" the only suffix supported?  Or can one use "us, ns, ms" for sub-second precision?  What about "m", "h", "d", etc?

Is there a formal spec here, and can we link it?  If not, I'm wary of using it.

Reply:
We're already using this format for the service config `timeout` field, so it's not really a new format that we're not already dealing with.

Reply:
Yes, I recently noticed it there as well, and I think this is a real problem.

If there's no spec for this format, I think we should either request one ASAP or define our own spec and use it instead (for this and `timeout`).

Reply:
We can't use something else, because we are depending on the default JSON conversion to make use of proto data internally.

While I agree that the proto documentation is not a rigorous formal spec, I think it's clear enough to understand.  The value will be encoded in seconds, with decimal digits to indicate fractional seconds, followed by a trailing 's'.  If other suffixes were supported, I think it would have said so.  So basically, when parsing the JSON, we strip off the trailing 's' from the string, and we accept any number of decimal digits up to 9.

If you want to ask the proto folks for a formal spec, I'm fine with that, but I'd be surprised if we run into any significant problems with the above.

Reply:
> We can't use something else, because we are depending on the default JSON conversion to make use of proto data internally.

But when we parse it from JSON in order to use it, we have to JSON-parse it as a string and then decode it -- you don't convert it back to a proto, right?  Right now we're using a Go library for that.  https://golang.org/pkg/time/#ParseDuration -- this _seems_ fine, but how can I know for sure if there isn't a spec?

> The value will be encoded in seconds, with decimal digits to indicate fractional seconds, followed by a trailing 's'.

That sounds perfect, let's use it.  Maybe add something about nano precision and the maximum possible value.

Otherwise our users will almost certainly try things like "50ms".  And that will work for Go clients with the parsing we're doing.  But then it might not work in C if it only supports seconds.  The failure modes at that point might be really hard to debug or even notice for the user, so I think we need to be as specific as possible.

Reply:
> > We can't use something else, because we are depending on the default JSON conversion to make use of proto data internally.
> 
> But when we parse it from JSON in order to use it, we have to JSON-parse it as a string and then decode it -- you don't convert it back to a proto, right?

Right, we don't convert it back to a proto.  What I meant was that internally, the service config is stored in proto form, and the google resolver needs to convert it from proto to JSON to feed it into gRPC core.  In order to avoid implementing custom conversion for each individual field in the google resolver (and duplicating it in each language), we want to just depend on the existing code in the protobuf library for converting the entire proto to JSON.  Thus, we need to use the format the protobuf code generates for a Duration field.  And the format for that is documented in https://developers.google.com/protocol-buffers/docs/proto3#json.

Like I said before, if you want to ask the protobuf team to give you something more formal than what's already written in that doc, that's fine with me.  But I personally don't see much ambiguity in what's already there.

> Right now we're using a Go library for that.  https://golang.org/pkg/time/#ParseDuration -- this _seems_ fine, but how can I know for sure if there isn't a spec?

It looks like that Go library will support a lot more formats that what the protobuf library will generate.  I'm not sure why you chose to use that library, but I don't think it's a very good choice, for exactly the reason you mentioned below: it will cause problems if we support certain formats in some languages but not others.  I suggest that you change grpc-go to accept only the format generated by the proto library.

> > The value will be encoded in seconds, with decimal digits to indicate fractional seconds, followed by a trailing 's'.
> 
> That sounds perfect, let's use it.

I think this is basically what's already written in the protobuf docs.  But if this wording makes you happy, I'm fine with it. :)

> Maybe add something about nano precision and the maximum possible value.

The protobuf doc already says the following:

"""
Generated output always contains 0, 3, 6, or 9 fractional digits, depending on required precision. Accepted are any fractional digits (also none) as long as they fit into nano-seconds precision.
"""

> Otherwise our users will almost certainly try things like "50ms".  And that will work for Go clients with the parsing we're doing.  But then it might not work in C if it only supports seconds.  The failure modes at that point might be really hard to debug or even notice for the user, so I think we need to be as specific as possible.

I don't know why you would think that we should support any suffix other than 's', since the protobuf doc doesn't say anything about that.  And I agree with you that it could cause problems if Go supports that and other languages don't.  So I think you should drop support for those suffixes in Go.

Reply:
> It looks like that Go library will support a lot more formats that what the protobuf library will generate. I'm not sure why you chose to use that library, but I don't think it's a very good choice

I agree, knowing what I know about the format now.  But as an example of what happens when things are under-specified, note that the official JSON->proto conversion library in Go uses `time.ParseDuration`, too:
https://github.com/golang/protobuf/blob/1643683e1b54a9e88ad26d98f81400c8c9d9f4f9/jsonpb/jsonpb.go#L760

> I think this is basically what's already written in the protobuf docs. But if this wording makes you happy, I'm fine with it. :)

Yes, this makes me happy.  Let's do it.  I can send a PR for the service config `timeout` field if you'd like.  How about this:

```
// The value is encoded in seconds, with optional decimal digits to indicate
// fractional seconds, followed by a trailing "s".  Leading and trailing zeros
// are permissible, and no zero is required before or after the decimal point,
// if present.  No more than nine digits after the decimal point are allowed.
//
// Examples: "1s", "01s", ".0120s".
```

Reply:
Works for me.